File: OOP_Kotlin/tema6/lab6_2.kt
/*
In Kotlin nu exista friend classes, insa variabilele declarate de tip val sau var daca sunt publice, se poate lucra in clasele derivate la fel ca intr-un friend class
Alta varianta nu am gasit de a inlocui friend
 */

/*
Am creat o interfata carte apoi trei clase derivate pentru a putea lucra cu variabilele din codul C++
in interfete toate functiile sunt abstracte
*/

interface Carte {
    var titlu: String
    var autor: String
    var anPublicare: Int
    var disp: Boolean

    fun afiseazaInfo()
    fun verifDisp()
}

class CarteTiparita(
    override var titlu: String = "",
    override var autor: String = "",
    override var anPublicare: Int = 0,
    override var disp: Boolean = true
) : Carte {
    override fun afiseazaInfo() {
        println("Titlu: $titlu")
        println("Autor: $autor")
        println("An Publicare: $anPublicare")
        println("Disponibilitate: ${if (disp) "Disponibila" else "Indisponibila"}")
    }

    override fun verifDisp() {
        println("Cartea $titlu este ${if (disp) "disponibila" else "indisponibila"}.")
    }
}

class CarteElectronica(
    override var titlu: String = "",
    override var autor: String = "",
    override var anPublicare: Int = 0,
    override var disp: Boolean = true,
    private var dimensiuneMB: Float = 0.0f
) : Carte {

    override fun afiseazaInfo() {
        println("Titlu: $titlu")
        println("Autor: $autor")
        println("An Publicare: $anPublicare")
        println("Disponibilitate: ${if (disp) "Disponibila" else "Indisponibila"}")
        println("Dimensiune (MB): $dimensiuneMB")
    }

    override fun verifDisp() {
        println("Cartea electronica $titlu este ${if (disp) "disponibila pentru descarcare" else "indisponibila"}.")
    }
}

class Biblioteca {
    fun modifAutor(c: Carte, a: String) {
        if (c.autor != a) {
            c.autor = a
            println("Autorul a fost modificat!")
        } else {
            println("Autorul este acelasi!")
        }
    }

    /*
    Am folosit companion object deoarece in codul C++ functiile urmatoare sunt statice
    Totodata, am vrut sa afisez informatiile cartilor fizice direct din clasa biblioteca, pentru asta sunt verifDispo si afiseazaInformatii
    Totusi, inca am nevoie de metode in clasa carteTiparita deoarece trebuie sa faca override la toate metodele din interfata Carte
    */

companion object {
        fun verifDispo(c: Carte) {
            println("Cartea ${c.titlu} este ${if (c.disp) "disponibila" else "indisponibila"}.")
        }

        fun afiseazaInformatii(c: Carte) {
            println("Titlu: ${c.titlu}")
            println("Autor: ${c.autor}")
            println("An Publicare: ${c.anPublicare}")
            println("Disponibilitate: ${if (c.disp) "Disponibila" else "Indisponibila"}")
        }
    }
}

fun modifTitlu(c: Carte, t: String) {
    if (c.titlu != t) {
        c.titlu = t
        println("Titlul cartii a fost modificat in $t")
    } else {
        println("Titlul este acelasi!")
    }
}

fun main() {
    val c1: Carte = CarteTiparita("Ion", "Liviu Rebreanu", 1920, true)
    val c2: Carte = CarteTiparita("Morometii", "Marin Preda", 1955, false)
    val ce1: Carte = CarteElectronica("1984", "George Orwell", 1949, true, 1.5f)

    val b = Biblioteca()

    Biblioteca.verifDispo(c1)
    Biblioteca.afiseazaInformatii(c1)
    println()

    modifTitlu(c1, "Ional")
    Biblioteca.verifDispo(c1)
    Biblioteca.afiseazaInformatii(c1)
    println()

    b.modifAutor(c2, "Marin Predut")
    Biblioteca.verifDispo(c2)
    Biblioteca.afiseazaInformatii(c2)
    println()

    ce1.verifDisp()
    ce1.afiseazaInfo()
}

File: OOP_C++/tema6/lab6_2.cpp
#include <iostream>
#include<string>
using namespace std;

class Carte{
private:
    string titlu;
    string autor;
    int anPublicare;
    bool disp;

public:
    Carte(string t="", string a="", int an=0, bool d=true): titlu(t), autor(a), anPublicare(an), disp(d) {}

    string getTitlu() const {
        return titlu;
    }

    string getAutor() const {
        return autor;
    }

    int getPublicare() const {
        return anPublicare;
    }

    bool getDisp() const {
        return disp;
    }

    virtual void modifTitlu(string t,string a) {}

    virtual void afiseazaInfo(){}



    friend class Raft;
    friend class Biblioteca;

};

class Raft: public Carte{
protected:
    int nr_raft;
public:
    Raft(string t="", string a="", int an=0, bool dis=true, int nr=0)
      : Carte(t, a, an, dis), nr_raft(nr) {}

    void cautaDisp(Carte &c) {
        if (c.disp == true) {
            cout<<"Cartea "<<c.titlu<<" este disponibila.\n";
        } else {
            cout<<"Cartea "<<c.titlu<<" nu este disponibila.\n";
        }
    }

    void afiseazaInfo() override{
        cout<<"Carte: "<<titlu<<", Autor: "<<autor<<", An Publicare: "<<anPublicare<<", Nr Raft: "<<nr_raft<<endl;

    }

    friend void modifRaft(Raft &r, int nr);

};

class Biblioteca:public Raft{
    public:
    Biblioteca():Raft("","",0,false,0) {}

    void modifTitlu(string t,string a) override {
        if (a==autor) {
            titlu = t;
            cout<<"Titlul cartii a fost modificat la "<<titlu<<".\n";
        }
        else {
            cout<<"Datele autorului nu corespund. Titlul nu a fost modificat.\n";
        }
    }

    void imprumutaCarte(Carte &c,Raft &r) {
        if (c.disp == true) {
            c.disp = false;
            cout<<"Ai imprumutat \n ";
            r.afiseazaInfo();
        } else {
            r.afiseazaInfo();
            cout<<endl<<"Nu este disponibila pentru imprumut.\n";
        }
    }


};

void modifRaft(Raft &r, int nr) {
    r.nr_raft = nr;
    cout<<"Numarul raftului a fost modificat la "<<r.nr_raft<<".\n";
}

int main() {
    Raft r1("Enigma Otiliei","George Calinescu",1938,true,5);
    Biblioteca b;
    Carte c1("Ion","Liviu Rebreanu",1920,true);
    Carte c2("Morometii","Marin Preda",1955,false);
    b.imprumutaCarte(c2,r1);
    b.modifTitlu("Ionelule","George Calinescu");


}